tests/tools/jira/test_jira_dryrun_client.py

```diff
@@ -0,0 +1,51 @@
+import os
+import pytest
+from unittest.mock import Mock
+from tools.jira_client.jira_api import CLOSED_STATUS
+from tools.jira_client.jira_dryrun_client import JiraDryRunClient
+
+
+def get_devnull_stream():
+    return open(os.devnull, "w")
+
+
+def get_mock_issue(key="foobar"):
+    issue = Mock()
+    issue.key = key
+    return issue
+
+
+class MockAttributeErrorClient:
+    def method_exists(self):
+        pass
+
+
+def test_should_throw_exception_when_calling_undefined_methods_and_real_jira_has_undefined_method_also():
+    mock_jira_client = MockAttributeErrorClient()
+    jira_dryrun_client = JiraDryRunClient(jira_client=mock_jira_client, dry_run_stream=get_devnull_stream())
+    with pytest.raises(AttributeError):
+        jira_dryrun_client.does_not_exist()
+
+
+def test_should_proxy_calls_to_real_jira_when_calling_undefined_methods():
+    mock_jira_client = Mock()
+    mock_jira_client.search_issues = Mock(return_value=[])
+    jira_dryrun_client = JiraDryRunClient(jira_client=mock_jira_client, dry_run_stream=get_devnull_stream())
```

Comment:
just a second, the dry-run client is able to proxy any call to jira client? how can we make sure it doesn't do write actions through those calls?

Reply:
It's the humans responsibility to add all the write methods to the JiraDryRun (just as we hide the write call behind an `if not dry run`, but automatically all read methods will work without the need to add them as boilerplate.

The alternative I see is to add _all_ the needed methods (read and write), and proxy the read ones manually.

The risk is that users will think this is a magic class and just using without adding write methods that might be used.

WDYT?

Reply:
Yeah, I guess it's not that easy to change
Ultimately we can define an interface with ``abc`` to make sure all required methods are implemented. but I get how it's too much code
I also thought about a read-only jira tokens, but it requires us to create a new user because there's no read-only scope we can switch to

All in all, I think it's fine as-is

Reply:
Got it, good point about the read token, we can keep it in mind for future iterations as I am aware this can potentially be dangerous.